### PR TITLE
#744: return Subscription instances from `pubsub.list_subscriptions()`

### PR DESCRIPTION
**Pull request by tseaver**
Includes regression tests for subscription creation, as well as fixes for bugs found via the regression tests.

See: #744

### CODE REVIEW

**Comment by tseaver** [comment]
@dhermes PTAL

**Comment by dhermes** [comment]
Generic question: Do we want `Connection` bound to objects? (I feel like we discussed this but I forgot.)

**Comment by tseaver** [comment]
`pubsub` follows the existing `storage` pattern.  I would be OK with updating both such that all API-triggering methods took an optional `connection` argument, falling back to the implicit one:  we would then drop the `connection` attribute altogether.

File: gcloud/pubsub/subscription.py [hunk 1]
```diff
@@ -74,9 +74,7 @@ def exists(self):
         """
         conn = self.topic.connection
         try:
-            conn.api_request(method='GET',
-                             path=self.path,
-                             query_params={'fields': 'name'})
```
Comment (dhermes) [inline gcloud/pubsub/subscription.py]: Why drop this? Since we don't use the payload, it minimizes the total size.

Comment (tseaver) [inline gcloud/pubsub/subscription.py]: The API doesn't support it.

Comment (tseaver) [inline gcloud/pubsub/subscription.py]: https://cloud.google.com/pubsub/reference/rest/v1beta2/projects/subscriptions/get

Comment (dhermes) [inline gcloud/pubsub/subscription.py]: https://developers.google.com/apis-explorer/#p/pubsub/v1beta2/pubsub.projects.subscriptions.get

has fields and the fields editor lists name.

What in particular isn't supported?

Comment (tseaver) [inline gcloud/pubsub/subscription.py]: Hmm, I think I ran into a 400 when passing `fields`.

Comment (dhermes) [inline gcloud/pubsub/subscription.py]: Can you reproduce?

Comment (tseaver) [inline gcloud/pubsub/subscription.py]: Maybe I mis-attributed another error:  b44aea94188f1b81ba64644afd2d335450a8b6e4 reverts that (and the related test change).

**Comment by dhermes** [comment]
LGTM. 6 commits a bit much for really two distinct ideas. Especially with 2 of the 6 just canceling out.

**Comment by tseaver** [comment]
@dhermes Squashed to two commits.  Unless you raise an objection, I will merge once Travis is green.

**Comment by dhermes** [comment]
Thanks for feeding my OCD. All systems go

**Pull request merged**